### PR TITLE
make round trip testing generic

### PR DESCRIPTION
**Pull request by deads2k**
RoundTrip testing is something associated with a scheme and everyone who writes an API will want to do it.  In the end, we should wire each API group separately in a test scheme and have them all call this general function.  Once `kubeadm` is out of the main scheme, we'll be able to remove the one really ugly hack.

@luxas @sttts @kubernetes/sig-apimachinery-pr-reviews @smarterclayton

### CODE REVIEW

**Comment by deads2k** [comment]
@kubernetes/sig-api-machinery-pr-reviews

File: staging/src/k8s.io/apimachinery/pkg/api/testing/roundtrip.go [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	runtimeserializer "k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/util/diff"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+var FuzzIters = flag.Int("fuzz-iters", 20, "How many fuzzing iterations to do.")
+
+// nonRoundTrippableTypes are kinds that are effectively reserved across all GroupVersions
```
Comment (sttts) [inline staging/src/k8s.io/apimachinery/pkg/api/testing/roundtrip.go]: globalNonRoundTrippableTypes

File: staging/src/k8s.io/apimachinery/pkg/api/testing/roundtrip.go [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+	// the schema by content type, rather than via kind/version included in each
+	// object.
+	"WatchEvent",
+	// ListOptions is now part of the meta group
+	"ListOptions",
+	// Delete options is only read in metav1
+	"DeleteOptions",
+)
+
+// RoundTripTypes applies the round-trip test to all round-trippable Kinds
```
Comment (sttts) [inline staging/src/k8s.io/apimachinery/pkg/api/testing/roundtrip.go]: RoundTripTypesWithoutProtobuf

File: staging/src/k8s.io/apimachinery/pkg/api/testing/roundtrip.go [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+	if !apiequality.Semantic.DeepEqual(original, object) {
+		t.Errorf("0: %v: encode altered the object, diff: %v", name, diff.ObjectReflectDiff(original, object))
+		return
+	}
+
+	// decode (deserialize) the encoded data back into an object
+	obj2, err := runtime.Decode(codec, data)
+	if err != nil {
+		t.Errorf("0: %v: %v\nCodec: %#v\nData: %s\nSource: %#v", name, err, codec, dataAsString(data), printer.Sprintf("%#v", object))
+		panic("failed")
```
Comment (sttts) [inline staging/src/k8s.io/apimachinery/pkg/api/testing/roundtrip.go]: return

File: staging/src/k8s.io/apimachinery/pkg/api/testing/roundtrip.go [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+		} else {
+			t.Errorf("%v: %v (%s)", name, err, printer.Sprintf("%#v", object))
+		}
+		return
+	}
+
+	// ensure that the deep copy is equal to the original; neither the deep
+	// copy or conversion should alter the object
+	// TODO eliminate this global
+	if !apiequality.Semantic.DeepEqual(original, object) {
```
Comment (sttts) [inline staging/src/k8s.io/apimachinery/pkg/api/testing/roundtrip.go]: actually this test is also fulfilled by any shallow copy. We should create 2 copies, fuzz one and compare the other copy with the original.

File: staging/src/k8s.io/apimachinery/pkg/api/testing/roundtrip.go [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+	obj2, err := runtime.Decode(codec, data)
+	if err != nil {
+		t.Errorf("0: %v: %v\nCodec: %#v\nData: %s\nSource: %#v", name, err, codec, dataAsString(data), printer.Sprintf("%#v", object))
+		panic("failed")
+	}
+
+	// ensure that the object produced from decoding the encoded data is equal
+	// to the original object
+	if !apiequality.Semantic.DeepEqual(original, obj2) {
+		t.Errorf("\n1: %v: diff: %v\nCodec: %#v\nSource:\n\n%#v\n\nEncoded:\n\n%s\n\nFinal:\n\n%#v", name, diff.ObjectReflectDiff(object, obj2), codec, printer.Sprintf("%#v", object), dataAsString(data), printer.Sprintf("%#v", obj2))
```
Comment (sttts) [inline staging/src/k8s.io/apimachinery/pkg/api/testing/roundtrip.go]: no need for \n

**Pull request merged**